File: 프로그래머스/lv1/신은혜/삼총사.java
class Solution {
    public int solution(int[] number) {
        int result=0;
        for(int i=0; i<(1<<number.length); i++){
            int sum=0;
            int cnt=0;
            for(int j=0; j<number.length; j++){
                if((i&(1<<j))!=0){
                    cnt++;
                    sum+=number[j];
                }
            }
            if(cnt==3&&sum==0)
                result++;         
        }
        return result;
    }
}

File: 프로그래머스/lv1/이하은/삼총사.kt
/*
n <= 13 
3개의 정수를 더해서 0이 되는 경우의 수 
*/
class Solution {
    lateinit var selected: BooleanArray
    val arr = IntArray(3) {0}
    var answer = 0
    
    fun solution(number: IntArray): Int {
        selected = BooleanArray(number.size) {false}
        
        dfs(0, 0, number)
        
        return answer
    }
    
    fun dfs(idx: Int, cnt: Int, number: IntArray){
        if(cnt == 3){
            if(arr.sum() == 0) answer++
            return
        }
        
        for(i in idx until number.size){
            if(!selected[i]){
                selected[i] = true
                arr[cnt] = number[i]
                dfs(i + 1, cnt + 1, number)
                selected[i] = false
            }
        }
    }
}